### PR TITLE
Add ability to set individual head QKV inputs via hook

### PR DESCRIPTION
**Pull request by epurdy**
Disabled by default to save GPU memory.

### CODE REVIEW

File: transformer_lens/HookedTransformer.py [hunk 1]
```diff
@@ -986,6 +986,12 @@ def set_use_attn_result(self, use_attn_result):
         """
         self.cfg.use_attn_result = use_attn_result
 
+    def set_use_attn_input(self, use_attn_input):
```
Comment (neelnanda-io) [inline transformer_lens/HookedTransformer.py:989]: set_use_attn_input_hooks seems like a better name (I should probably change use_attn_result too lol)

File: transformer_lens/components.py [hunk 2]
```diff
@@ -298,7 +301,9 @@ def __init__(
         # See HookedTransformerConfig for more details.
         if self.cfg.positional_embedding_type == "shortformer":
             # This tracks the input to the keys and queries, which is resid_pre + pos_embeds
-            self.hook_attn_input = HookPoint()  # [batch, pos, d_model]
+            # The dimension of these depends on whether cfg.use_attn_input is False (1st) or True (2nd)
+            self.hook_attn_input_q = HookPoint()  # [batch, pos, d_model] OR [batch, pos, head_index, d_model]
```
Comment (neelnanda-io) [inline transformer_lens/components.py:305]: Oh man, the shortformer bits of this are really ugly. I weakly lean towards just having this feature only apply to non shortformer positional embeddings...?

File: transformer_lens/components.py [hunk 3]
```diff
@@ -350,38 +355,95 @@ def forward(
         past_kv_cache_entry is an optional entry of past keys and values for this layer, only relevant if generating text. Defaults to None
 
         """
+        # If user requests it, make a hookpoint with separate inputs for each head.
+        # This allows the user to change the inputs to different heads independently.
+        if self.cfg.use_attn_input:
+            input_to_q = self.hook_input_q(
+                resid_pre.unsqueeze(2).expand(-1, -1, self.cfg.n_heads, -1).clone()
```
Comment (neelnanda-io) [inline transformer_lens/components.py:362]: Please rewrite to einops

`einops.repeat(resid_pre, "batch pos d_model -> batch pos head_index d_model", head_index=self.cfg.n_heads)` should work and IMO is much more readable

File: transformer_lens/components.py [hunk 4]
```diff
@@ -287,6 +287,9 @@ def __init__(
         if self.cfg.scale_attn_by_inverse_layer_idx:
             self.attn_scale *= self.layer_id + 1
 
+        self.hook_input_q = HookPoint()  # [batch, pos, head_index, d_model]
```
Comment (neelnanda-io) [inline transformer_lens/components.py:290]: Add comment documenting them

File: transformer_lens/components.py [hunk 5]
```diff
@@ -350,38 +355,95 @@ def forward(
...
+            input_to_v = self.hook_input_v(
+                resid_pre.unsqueeze(2).expand(-1, -1, self.cfg.n_heads, -1).clone()
+            )
+        else:
+            input_to_q = input_to_k = input_to_v = resid_pre
+
         if self.cfg.positional_embedding_type in ["standard", "rotary"]:
             # Normal attention
-            q = self.hook_q(
+            if self.cfg.use_attn_input:
```
Comment (neelnanda-io) [inline transformer_lens/components.py:377]: Add comment documenting why this if statement is there

**Review by neelnanda-io — CHANGES_REQUESTED** [review]
Thanks for this! Overall looks fairly good, I left some comments for cosmetic changes.

A galaxy brained idea - you *can* broadcast a tensor up to a larger, repeated shape without changing the total GPU memory used, since it just changes the striding on the same underslying memory (the bonus axis has stride=0). So maybe we should implement it like that, so we don't need to have the branching statements for q, k and v. OTOH this is too clever and will probably break something rip.

But yeah, I'm overall torn between this being a useful feature and making the code too ugly, though overall I lean towards this being worth it.

**Comment by ArthurConmy** [comment]
I implemented this here https://github.com/ArthurConmy/Automatic-Circuit-Discovery in einops. I'm happy to take this if wanted @epurdy ?

**Comment by ArthurConmy** [comment]
Also, I am in favour of the galaxy brain solution. Note this could apply to `use_attn_result` too: we would just broadcast up `attn_out` to have an extra dimension with `einops.repeat`

(and then if we `set_use_attn_results`, we'd clone the tensor so that there wasn't a stride 0 element in there)

I can explain, if this was unclear

**Comment by epurdy** [comment]
> I implemented this here https://github.com/ArthurConmy/Automatic-Circuit-Discovery in einops. I'm happy to take this if wanted @epurdy ?

Sure, @ArthurConmy, that would be great.

**Comment by neelnanda-io** [comment]
Sounds great Arthur, it would be good to include a test that the various
acts are actually stride zero in the expected way, and that you can't use
the hook without setting use_... in the relevant way

**Comment by ArthurConmy** [comment]
See https://github.com/neelnanda-io/TransformerLens/pull/158

**Pull request closed without merging**